### PR TITLE
Added LogAnalyzer for checking changes on Satellite log files

### LINKED ISSUE #447: Log analyzing Tool while Upgrade Running
Implement / Come up with Log analyzing tool while Upgrade in progess/running, which will:
1. Alert if Error/Exception observed in bg logs
2. Save/Collect all the logs for debugging the failures later

### PR DESCRIPTION
**Pull request by renzon**
close #447

Context Manager was added to upgrade task

### CODE REVIEW

**Comment by renzon** [comment]
By now I am only printing lines appended to log files. On the upgrade process I got no errors during upgrade, so I could process errors. We can start simple like this, just checking what changed on log files and enhancing process logic once we face errors. What you guys think? @jyejare @omaciel @sghai.

In time, I am just printing the changes right now, so logs could be seen on Jenkins builds.

**Comment by sghai** [comment]
@renzon : I think its a very good start to print the logs appended to log files. However, next step would be to give us a  summary like:

3 errors appeared in /var/log/messages during `upgrade_task` or just 3 errors appeared in '/var/log/messages' and later may be one liner from those errors/exceptions.

@jyejare @omaciel : thoughts ?

File: automation_tools/satellite6/log.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from fabric.operations import run
```
Comment (jyejare) [inline automation_tools/satellite6/log.py]: run and execute can be imported from fabric,api and hence club into one line

File: automation_tools/satellite6/log.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+from fabric.tasks import execute
```
Comment (jyejare) [inline automation_tools/satellite6/log.py]: update this as per above comment

**Comment by jyejare** [comment]
@sghai :+1:  for what you said !

File: automation_tools/satellite6/log.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    def _update_log_files_state(self):
+        """Update log_dct with adding delta from current number of lines of
+        each
+        item and the last provided by dct. Ex:
+
+        So this method can be used to check how many lines were appended on
+        a file
+        during some processes and used to tail them. If log dct is None a
+        new dict
+        is created
```
Comment (sghai) [inline automation_tools/satellite6/log.py]: nitpick: I feel these doc strings required a little bit formatting.

File: automation_tools/satellite6/log.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+                    analyze(log_file, content)
+                else:
+                    print('### No changes in %s' % log_file)
+
+        execute(fetch_appended_log_lines, host=self.sat_host)
+
+    def _update_log_files_state(self):
+        """Update log_dct with adding delta from current number of lines of
+        each
+        item and the last provided by dct. Ex:
```
Comment (jyejare) [inline automation_tools/satellite6/log.py]: The line is not appended properly. Please append all the lines in this functions docstrings.

File: automation_tools/satellite6/log.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        def fetch_appended_log_lines():
+            for log_file, lines_appended in self.log_state.items():
+                if lines_appended > 0:
+                    content = run(
+                        'tail -n {} {}'.format(lines_appended, log_file),
+                        quiet=True
+                    )
+                    analyze(log_file, content)
+                else:
+                    print('### No changes in %s' % log_file)
```
Comment (jyejare) [inline automation_tools/satellite6/log.py]: I think this is not rwquired. We just want log file changes if there are. So i am ok to remove else loop.

File: automation_tools/satellite6/log.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+        new dict
+        is created
+
+        :param sat_host: str with satellite hostname
+        :param log_dct: dict to be updated
+        """
+
+        def get_line_count():
+            for log_file, old_value in self.log_state.items():
+                current_value = int(
```
Comment (jyejare) [inline automation_tools/satellite6/log.py]: As per the code, I understood:
When upgrade will start with the context manager, initially, the log_state[log_file] = current_value(an example that in some logfile there will 500 line already then this current value will become 500) - old_value(which will be 0) = 500

Then, that 500 number will be sent to fetch_appended_log_files() def and the content = tail -n 500 <logfilename>, which will have last 500 line in that file, which will be sent to analyze() def to print which wiil really be mess an unecessary logs regardless to upgrade/current process.

I may understood it wrong, please correct me i m wrong.

File: automation_tools/satellite6/upgrade/__init__.py [hunk 7]
```diff
@@ -167,17 +169,20 @@ def product_upgrade(product):
...
-                execute(satellite6_capsule_upgrade, cap_host, host=cap_host)
-                # Generate foreman debug on capsule after upgrade
-                execute(
-                    foreman_debug,
-                    'capsule_{}'.format(cap_host),
-                    host=cap_host)
-        if product == 'client':
-            satellite6_client_upgrade('rhel6', clients6)
-            satellite6_client_upgrade('rhel7', clients7)
+        with LogAnalyzer(sat_host):
```
Comment (jyejare) [inline automation_tools/satellite6/upgrade/__init__.py]: Repeat the LogAnalyzer context manager on sat_host(which you are doing already) and on cap_host.
I will take care about clients as they are situated in Docker and the whole process is bit complex.

**Comment by renzon** [comment]
@sghai: Some log files follow log convention putting log level (INFO, ERROR...) on them  so I will change to analyse errors and warnings based those statuses. 

But some files, doesn't have this format. Check the pattern on production.log:

```console
2016-11-23 05:41:55 [app] [I]   Rendered api/v2/reports/create.json.rabl (5.5ms)
2016-11-23 05:41:55 [app] [I] Completed 201 Created in 61ms (Views: 5.7ms | ActiveRecord: 24.0ms)
```
So if I can somehow (doc, talking to dev) know how a error message looks like on this kind of file, I can improve error handling.

File: automation_tools/satellite6/log.py (hunk 6 above)
Comment (renzon) [inline automation_tools/satellite6/log.py]: @jyejare in fact I update the dict with the delta only. So at first I call update log_state and it gets a value of 500. Suppose upgrade append just one line. So current_value gets a value of 501 and dict is updated with 501 - 500 (line 103). Thus tail will be called only with 1, which is exactly the line appended during process. That was my main concern while thinking on this solution, not having to parse and print the entire log file ;)

**Comment by sghai** [comment]
@renzon : I'll share some errors as an example that we found during earlier upgrades if that helps.

**Comment by sghai** [comment]
A few very common errors:

`candlepin.log:`
- javax.persistence.PersistenceException: unexpected error when rollbacking

production.log
- Could not find the inverse association for repository
- PG::Error: ERROR:  update or delete on table
- NoMethodError: undefined method
- usr/sbin/foreman-rake db:migrate returned 1 instead of one of [0]
- /usr/sbin/foreman-rake db:seed returned 1 instead of one of [0]
- [E] undefined method `finished' for nil:NilClass (NoMethodError)
- [ProxyAPI::ProxyException]: Unable to detect
- (ActiveModel::MissingAttributeError)
- ActionView::Template::Error (undefined method
- undefined method `cp_config' for
- An error has occurred, this and all later migrations canceled
- undefined method `import_data' for nil:NilClass (NoMethodError)
- undefined method `[]' for nil:NilClass (NoMethodError)

These errors are common to various different issues. I'll see more and update here.

File: automation_tools/satellite6/log.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+        execute(get_line_count, host=self.host)
+
+
+def analyze(log_file, content):
+    """Analyzes appended content from a log file. For now it is only
+    checking for ERROR status on regular log files.
+    """
+    _print_wrapper('### Analyzing %s:' % log_file)
+    errors = sum(map(lambda _: 1, re.finditer('ERROR', content)))
```
Comment (jyejare) [inline automation_tools/satellite6/log.py]: @renzon :+1:  for using map, RE and sum for finding the number of total Errors. What a convenient way to use them . You are the real python champ.

Comment (jyejare) [inline automation_tools/satellite6/log.py]: You could have also don this :
```
errors = len(re.findall('ERROR', content))
```

File: automation_tools/satellite6/upgrade/__init__.py [hunk 9]
```diff
@@ -3,9 +3,9 @@
 Many commands are affected by environment variables. Unless stated otherwise,
 all environment variables are required.
 """
-import os
 import sys
 
+import os
```
Comment (jyejare) [inline automation_tools/satellite6/upgrade/__init__.py]: This should go up and above ```import sys``` alphabetically.

**Comment by renzon** [comment]
@jyejare @sghai I changed the approach. Now I numbered lines on report and instead of showing only error count, I show error lines themselves.

**Comment by renzon** [comment]
I have created also a list of tokens which can define an error. So you evolve it during our explorations: 

https://github.com/SatelliteQE/automation-tools/pull/459/files#diff-0bc53cb6a415522627ad706a40e0a1aaR107

File: automation_tools/satellite6/log.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    _print_wrapper('### Analyzing %s:' % log_file)
+    _print_wrapper('## Errors found:')
+    content_lines = content.split('\n')
+    error_tokens = (
+        'ERROR',
+        'SEVERE',
+        'EXCEPTION',
+        r'returned 1 instead of one of \[0\]',
+        'Could not find the inverse association for repository',
+        'undefined method'
```
Comment (sghai) [inline automation_tools/satellite6/log.py]: 👍

File: tests/test_log.py [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+10: /usr/sbin/foreman-rake db:seed returned 1 instead of one of [0]
+11: [E] undefined method `finished' for nil:NilClass (NoMethodError)
+12: [ProxyAPI::ProxyException]: Unable to detect
+13: (ActiveModel::MissingAttributeError)
+14: ActionView::Template::Error (undefined method
+15: undefined method `cp_config' for
+16: An error has occurred, this and all later migrations canceled
+17: undefined method `import_data' for nil:NilClass (NoMethodError)
+18: undefined method `[]' for nil:NilClass (NoMethodError)'''
+
```
Comment (sghai) [inline tests/test_log.py]: 👍

**Comment by renzon** [comment]
I have created also a list of tokens which can define an error. So you evolve it during our explorations: 

https://github.com/SatelliteQE/automation-tools/pull/459/files#diff-0bc53cb6a415522627ad706a40e0a1aaR107

File: automation_tools/satellite6/log.py [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+    #### Analyzing logs of host root@sathost.redhat.com
+    [root@sathost.redhat.com] Executing task 'get_line_count'
+    [root@sathost.redhat.com] Executing task 'fetch_appended_log_lines'
+
+    """
+
+    def __init__(self, sat_host):
+        """Initializes context manager with Satellite hostname
+
+        :param sat_host: str
```
Comment (jyejare) [inline automation_tools/satellite6/log.py]: Description of parameter is missing

File: automation_tools/satellite6/log.py [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+
+    def _update_log_files_state(self):
+        """Update log_dct with adding delta from current number of lines of
+        each item and the last provided by dct.
+
+        So this method can be used to check how many lines were appended on
+        a file during some processes and used to tail them. If log_state is
+        None a new dict is created with initial values been 0 (zero).
+
+        :param host: str with satellite hostname
```
Comment (jyejare) [inline automation_tools/satellite6/log.py]: Parameter doesnt exist anymore

File: automation_tools/satellite6/log.py [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+        :param log_dct: dict to be updated
```
Comment (jyejare) [inline automation_tools/satellite6/log.py]: Parameter doesnt exist anymore

File: automation_tools/satellite6/log.py [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+                else:
+                    self.log_state[log_file] = current_value - old_value
+
+        execute(get_line_count, host=self.host)
+
+
+def analyze(log_file, content):
+    """Analyzes appended content from a log file. For now it is only
+    checking for ERROR status on regular log files.
+    """
```
Comment (jyejare) [inline automation_tools/satellite6/log.py]: parameters description is missing

File: automation_tools/satellite6/log.py [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+
+    def print_lines(lines_enumeration):
+        """print lines with numbers"""
+        for i, line in lines_enumeration:
+            _print_wrapper('{}: {}'.format(i, line))
+
+    _print_wrapper('### Analyzing %s:' % log_file)
+    _print_wrapper('## Errors found:')
+    content_lines = content.split('\n')
+    error_tokens = (
```
Comment (jyejare) [inline automation_tools/satellite6/log.py]: I think we should move this somewhere else, the list will increase in future. And its better to have its own place somewhere,

**Comment by renzon** [comment]
@jyejare I moved tokens to a constant at the end of module, this way we can grow it without messing with analyze code. Do you think is enough or would like a module only for this constant?

**Comment by jyejare** [comment]
@renzon , I wiil be happy if you move that to the top that module. That way another person can find it easily to update. But thats ok. Its a non blocking comment.

**Comment by renzon** [comment]
@jyejare I have checked /SatelliteQE/robottelo/issues#1988 on robottelo. Right now I am not failing the build because of error (returning 1 to system). Let me know if want me to fail it if some error is found on logs.

**Comment by jyejare** [comment]
@renzon , No , I am not in favor of Failing the build. Because those errors/exception can be related to pulp data migration and not the installer ones. As of now we just want to list the errors upfront and dont want manually check the logs. What do you think @sghai ?

**Comment by sghai** [comment]
@renzon @jyejare : so Idea here is when we are running upgrade via installer and some error appears but its not related to installer/installer.log rather related to some log file then:

1) we would want the installer should keep on running
2) once its finished, it should throw a summary of errors found from selected log files.

In case, error appears while running the installer and it belongs to installer.log then build should auto failed.

**Comment by renzon** [comment]
Ok, so it is not related to this issues, once we are not analyzing installer.log itself. So I think current version is good to merge after you guys review it.

File: automation_tools/satellite6/upgrade/__init__.py (hunk 7 above)
Comment (jyejare) [inline automation_tools/satellite6/upgrade/__init__.py]: What happened to this? Please run Log Analyzer on Capsule host as well !!

**Comment by jyejare** [comment]
ACK pending comment

**Comment by renzon** [comment]
@jyejare: I am log analyzer to capsules, check this line: https://github.com/SatelliteQE/automation-tools/pull/459/files#diff-b6179f964acf95c58503d1f68b443781R179. Code will not fail if there are missing log files, it will only skyp each one that is not present. But I am not sure if capsule log files are the same as satellite.

**Comment by jyejare** [comment]
@renzon , Ohh Now I see you added the log analyzer to Capsules. Thanks !!

**Comment by renzon** [comment]
@sghai can you review it again?

**Comment by sghai** [comment]
@renzon : PR looks great.. however for capsule a few log files won't be applicable. 
Installer file should be /var/log/foreman-installer/capsule.log

below files are not applicable for capsule:
- /var/log/foreman/production.log
- /var/log/candlepin/candlepin.log
- /var/log/tomcat/catalina.out
- /var/log/mongodb/mongodb.log

Only these files are applicable for capsule:
- /var/log/foreman-proxy/proxy.log
- /var/log/messages'

**Comment by jyejare** [comment]
@renzon , @sghai , We also want this tool supported for Satellite 6.0 and 6.1, Where the log files name were different as below:
```
Satellite:
/var/log/satellite-installer/satellite-installer.log

Capsule:
/var/log/capsule-installer/capsule-installer.log
```
So please ass these file names in our list. Anyhow these flies wont be available in 6.2 and 6.3 and as per our code the files that wont exists will be skipped. So no worries I believe.

**Comment by renzon** [comment]
@sghai @jyejare I just added log files: https://github.com/renzon/automation-tools/blob/b78938a8fa78eb4d7645ebd74e2b8830f4b33e3f/automation_tools/satellite6/log.py#L20-L21. Even with old files I got an error when testing upgrade where some of them were not available. Probably this happened because I had a fresh installation and no log were yet written. So I just made the code ignore missing files: https://github.com/renzon/automation-tools/blob/b78938a8fa78eb4d7645ebd74e2b8830f4b33e3f/tests/test_log.py#L133-L133.

**Comment by sghai** [comment]
@jyejare @renzon : I think PR is good to merge. Please feel free to raise PRs for other changes/improvements later.

File: automation_tools/satellite6/log.py [hunk 17]
```diff
@@ -0,0 +1,139 @@
+# -*- coding: utf-8 -*-
```
Comment (elyezer) [inline automation_tools/satellite6/log.py]: According to the PEP 263 [1] the proper format is `# coding=<encoding name>`. The other formats `# -*- coding: <encoding name> -*-` and `# vim: set fileencoding=<encoding name> :` are editor specific and supported by Python as well.

This is not a blocker but a curiosity, I've started using the Python specific format on my files and wanted to share this info.

[1] https://www.python.org/dev/peps/pep-0263/

**Comment by renzon** [comment]
@elyezer To be honest my Pycharm is configured to add coding reader and unicode literals on every file almost since I started to learn the language and not knowing much of the reason of it.

After that I discovered that any coding <encode> format would work, but "most" people use '# -*- coding: <encoding name> -*-' because it is specific to Emacs, editor that most Python core developers use (urban legend maybe).

But I never read the official PEP, thanks for sharing @elyezer. I going to update my Pycharm settings.

**Comment by elyezer** [comment]
> But I never read the official PEP, thanks for sharing @elyezer. I going to update my Pycharm settings.

As I said this is more informative than a real issue. Also it was same here, I didn't know about that PEP but took some time some time ago to read it.

Feel free to use whatever is ok for you :smile:

**Comment by sghai** [comment]
Merging now..

**Pull request merged**